cartoframes/credentials.py

```diff
@@ -1,208 +1,127 @@
-    """
-    stored = _check_overwrite('base_url', overwrite)
-    stored['base_url'] = base_url
-    with open(_TARGETPATH, 'w') as outfile:
-        json.dump(stored, outfile)
-    return _TARGETPATH
-
-
-def set_api_key(key, overwrite=False):
-    """Save the CARTO API key so that users can access it via
-    `cartoframes.credentials.api_key()`. This lets users bind an API key to a
-    given installation.
-
-    Args:
-        key (str): CARTO API key
-        overwrite (bool): Whether to overwrite existing key. Defaults to False.
-
-    Returns:
-        str: Path to the location of the API key file. Raises FileError if
-        overwriting is prohibited and the file exists.
-    """
-    stored = _check_overwrite('api_key', overwrite)
-    stored['api_key'] = key
-    with open(_TARGETPATH, 'w') as outfile:
-        json.dump(stored, outfile)
-    return _TARGETPATH
-
-
-def api_key():
-    """Returns stored API key that was set with `cartoframes.credentials.set_key`
-
-    Returns:
-        str: CARTO API key
-    """
-    return _load_api_key()
-
-
-def base_url():
-    """Returns stored base_url that was set with
-    `cartoframes.credentials.set_base_url`
-
-    Returns:
-        str: CARTO API base URL
-    """
-    return _load_base_url()
-
-
-def credentials():
-    """Returns stored credentials that was set with
-    `cartoframes.credentials.set_credentials`
-
-    Returns:
-        dict: Dictionary containing CARTO API base_url and API key.
-    """
-
-    return _load_creds()
-
-
-def _load_creds():
-    """Load both URL and api_key from the credentials file
-
-    Returns:
-        dict: Dictionary containing keys `base_url` and `api_key`, values are
-        strings. If the credential is not set, the value may be an empty
-        string.
-    """
-    try:
-        with open(_TARGETPATH, 'r') as f:
+# import warnings
+import appdirs
+
+_USER_CONFIG_DIR = appdirs.user_config_dir('cartoframes')
+_DEFAULT_PATH = os.path.join(_USER_CONFIG_DIR,
+                             'cartocreds.json')
+
+
+def create_config_dir():
+    """create directory if not exists"""
+    if not os.path.exists(_USER_CONFIG_DIR):
+        os.makedirs(_USER_CONFIG_DIR)
+
+
+class Credentials(object):
+    """Credentials class for managing and storing user CARTO credentials"""
+    def __init__(self, key=None, username=None, base_url=None, cred_file=None):
+        if (key and username) or (key and base_url):
+            self._key = key
+            self._username = username
+            if base_url:
+                self._base_url = base_url
+            else:
+                self._base_url = 'https://{}.carto.com/'.format(self._username)
+        elif cred_file:
+            self._retrieve(cred_file)
+        else:
+            try:
+                creds = json.loads(open(_DEFAULT_PATH).read())
+                self._key = creds['key']
+                self._base_url = creds['base_url']
+                self._username = creds.get('username')
```

Comment:
This should probably just be `self._retrieve()`, maybe with some error handling.

Reply:
Totally, good catch!